### ISSUE TITLE
API design, ideas, thoughts, constructive critics - [PostJSON / LimonJS collaboration]

### ISSUE THREAD

**Issue by tunnckoCore** [cssinjs/jss#227]
Started from one thread in `cssx` - https://github.com/krasimir/cssx/issues/2, and my already started work on https://github.com/postjson/postjson, https://github.com/limonjs/limon and upcoming "pluggable" parser.
# JSS Thoughts

My first thoughts are that jss have not good api, too complex and wrong by design - wrong plugin signature. 
1. plugin should return "the new style" - actually, just some "final" style, object.
2. plugin gives only single "rule" object - which in first hand is complex and too deep, in second hand you are forced to access options and jss itself through this "rule" object coming.
3. a lot better and es6 ready (meaning use of arrow functions) will be the signature of plugins to be something more easy and meaningful like

``` js
jss.use(app => {
  return (key, value, rule) => {
    // stuff here, where:
    // `this` if not arrow function will be the `rule` object,
    // otherwise you will have access to it through `app` - somelike "current rule" pointer 
    // in that context we can have useful metadata info for the rule like: name, className and etc 
    // `value` is "style" object as it is currently, meaning the value of selector, containing the prop:val pairs
    // `key` is the "selectorText" string
  }
})
```

this way you have anything not so deep and not so complex, and it won't be a problem each rule still to have own "options" and meta stuff, like "name", "originalStyle" (which can be cloned `value` to be able to transform/modify `rule` directly) and etc. Even more, plugins will be able to extend `jss` api itself when not returning function.
# The Plugin API

All this plugin architecture can be done using http://npm.im/use, but currently (and maybe won't in future) not support passing more than one argument to `.run` method (meaning, what you pass to run is then passed to function return from plugin) and it must be object.
So, because of that I reimplemented it in [limonjs/limon/index.js#L49-L69](https://github.com/limonjs/limon/blob/master/index.js#L49-L69).
This way it won't be needed to "return the new/modified rule" from each plugin, which will be better.
# AST a.k.a. the "JSS JSON"

Actualy, jss don't have AST, and okey because of that you calling it "JSS JSON", but when rethinking it these weeks, we really dont need AST. PostJSON will allow access to (and produce some) AST, but mainly, plugins will expose to you `key, value, obj` signature. I also thinking they can be separate things (two type of plugins) - AST plugins, and default plugins.
In anyway, you don't have AST, you have some deep and complex (and circular) objects with metadata and some state - it can be rethinked and can be flattened.
PostJSON future plugin example

``` js
var postjson = require('postjson')

postjson.use(function (app) {
  // where `app` and `this` context are the postjson instance
  // if you return function it will be called
  // with different arguments, but again context of it will be `app`
  return function (key, value, obj) {
    console.log(obj[key] === value)
  }
})
```

Mainly it will work `.use().use().use().process(jsonStringifiedObj / orRawObject)`, secondly it will have "api extending plugin" which adds `.parse` method which will produce AST if user want to work with AST and not with main plugin signature. If this "api extending plugin" is used it maybe should change the signature of inner returned function - for example it should be

``` js
var postjson = require('postjson')

postjson.use(astPlugin).use(function (app) {
  // where `app` and `this` context are the postjson instance
  return function (ast) {
    console.log(ast)
  }
})
```

or preserve original signature and add the ast to `app/this` instance.
# JSS API

Talking for the `.createRule` and `.createStyleSheet` and etc. They also can be done better and this will be able when/if we change the plugins api. They also produce deep and complex objects.
What I noticed is that `.createRule` have `name, val` signature, where `name` is the clean selector called "name", which in turn actually represents "key" in JSON object. And where `val` is that "key"s value and more.. it is not limited to be always object, which is strange - it can be string and it is then apply to "originalStyle" and "style" properties. This is kinda bug. Maybe it is in that what to also be able to produce "prop: value" pairs, not only "real rules".
Example which produces identical objects if not 1:1 ->

``` js
var jss = require('jss')
var style = {
  'a > b:hover': {
    color: 'red'
  }
}

var sheet = jss.createStyleSheet(style)
var rule = jss.createRule('a > b:hover', {
  color: 'red'
})
console.log(sheet)
console.log('========')
console.log(rule)
```

Cheers,
Charlike

**Comment by tunnckoCore** [comment]
All in all, writing this and reviewing some things around i don't think to deal with AST in postjson, because I'm working on lower-level things - Limon pluggable lexer, which will allow me to create new CSS/SemVer/JSON lexers, parsers, stringifiers. And if users want AST they will use them, not postjson or anything like it. So, I'm trying to separate the things further and further, going deeper and deeper in low-levels.

**Comment by kof** [comment]
1. To clarify the jss api:

  I am not sure what you mean we can do better about jss api.

  - createRule name param is called name, because it can be a key or a selector, depending on the style sheet option `named` If named is true, which it is by default, `name` is a key. Of couse I could have called it a key, not sure if this makes it better.
  - createStyleSheet produces a style sheet, it is not the same as just producing a rule instance.

2. I am not sure why you think plugins api with signature `use(key, value, rule)` is much better

  I think your assumption is that plugins just modify the keys and values. For jss plugins it is often required to access the `rule` and `sheet`.

3. I think to understand the requirements for plugins api in jss, you need to know everything they need to do.  

I think this is something that I should add to the documentation of plugins api. And I agree there should be a better way, right now I know more or less all cases and can think of something better.

A plugin currently can:

  - rename a style key
  - rewrite a style value
  - remove a style key
  - add a style key
  - set/get rule.selector  
  - add a new rule to the sheet of the passed rule
  - get rule.type

**Comment by kof** [comment]
For me, good questions are:

-  should we separate plugins into just style modifications and more powerful with access to rule and more
-  should plugins always return a new style or be able to modify it

Also I would need to create benchmarks #48  to ensure this changes result in better performance, not worse.

**Comment by kof** [comment]
(key, value, rule) signature where key is selector or name, value is style makes it harder to understand what is what. First of all, plugins do not care a lot about the key. In named sheets key is users name for the rule. It doesn't matter for the plugin. Plugin may care about `rule.selector`, but it is not the same as key. Also it is not obvious to call `value` the `style object`.

**Comment by kof** [comment]
All in all I understand your motivation about plugins api. However your suggestion doesn't fit the requirements plugins API have, because you lack the knowledge of what is what here and how it is used. Take your time, try to understand the current abstraction, try to build a prototype of what you think is good. Come back, I am def. interested in your feedback.

**Comment by tunnckoCore** [comment]
> Take your time, try to understand the current abstraction, try to build a prototype of what you think is good. Come back, I am def. interested in your feedback.

Yea, I'll be around, working on similar things.

> style makes it harder to understand what is what.

It's not harder, because users should think in context of json. This all is just actually, right? So if you just say to your user "hey jss just work with json, accepts json, think for it like json" and promote this thinking, users easily will understand what is what - it is natural and it is visible - "my selector (or property if nested) is key of some object" and "my value is value", there's no easier, natural and expected way.

> A plugin currently can:

With proposed api this still be able to do it - just will be more flattened (clear if you want) not so deep and complex. You still have access to the rule - as you see, it is third argument. Actually, maybe make more sense to call it "sheet" or something like that, because my thinking for `obj` in the `key, value, obj` signature is that `obj` is some outer object - for example, say you have `this.rules`, so `this.rules[key]` should be equal to that `obj`. And `key` is rule name, selector, className or whatever you call it. The `value` is the "style" property (btw, i just have some bad feeling and don't like to call such thing "style", maybe declarations make more sense or at least it is more CSS-ish language, they are properties or as I call them in low-level "key:value pairs"). Style.. style for me is the document or at least single rule, but the whole rule object, not the selector value/body - for example `{selector: {width: 123, color: red}}`.

I understand that you separate these 3 things in your mind or for some reason and use-case. But.
But this is not needed and it is not useful. It adds more complexity and confusion, than it solves something, i'm 99% sure. 
Why? I'll answer you why. Because, for example, you mention the `named` option - okey. I played with it and it's awful behavior - completely awful. I can't think what it can solves. So.. why?
Because if we have this example

```js
// Create your style.
const style = {
  'div1 > a:hover': {
    color: 'red'
  }
}
jss.use(function (foo) {
  console.log('something:', foo)
})

// var sheet = jss.createStyleSheet(style, {named: false})
var rule = jss.createRule('div2 > a:hover', {
  color: 'red'
}, {named: false})
```
if `named` is true (as default), it's seems okey, I like it, it is correct. But. But it dissappears if you pass `named:false` - first, that's not okey to me. second there's no guarantee what is the selector when it named is false - "className" is changed to empty string, "selectorText" is changed to more meaningful thing - okey, seems like a real selector what I passed in. But. But when awfulness comes to these two properties when `named:true` - they are appended with the `id` (and there selectorText is also prefixed with some strange dot, why?), producing absolutely not correct and not valid "real selector" and not valid class name.You can't pass selectorText neither className to html `class="foo"` attribute neither it does something as rendered selector, neither as for example "jquery selector". Because "selectorText" is `.div2 > a:hover--fsdf34rjlk34h53----da-sdas-0-da023`, okey `.div2 > a:hover--jss-0-0` - but you can't pass to html class attribute, you can't pass it to some selector engine and you can't produce valid css with it - does nothing. All that when `named: false`.

Okey, you will say, "that's why it is true by default" - yes, but when option is true it is completely different view, you should then use `name` not `selectorText/className`.

> because you lack the knowledge of what is what here

yea you're right, absolutely. I reviewing deeper and deeper, and it goes harder and harder, more complex and complex. Unnecessary complex and non-logical, absolutely confusing.

***

I can continue and I'll continue to investigate and learn more and more, and will  work on the my parts. And I'm sure all that is small part of all (from first look) and I don't want to think how much complex, confusing and non-logical can/will go.

***

Really sorry if this is offensive. It is just first look and my thoughts. I'm not attacking you, i'm not offensive. I just can't realize how so complexity is possible.

Cheers,
Charlike! :beers:

**Comment by kof** [comment]
> It's not harder, because users should think in context of json

I know what you mean, however still even if we call it right: `(name, style, rule)` we still have the rule object which is still deep and complex, like you said.

We can also have this signature: `name, style, sheet`. The rule instance can be obtained by `sheet.getRule(name)`.

Still no less complexity here.

> style for me is the document

It is called `style` because [dom implementation](https://developer.mozilla.org/de/docs/Web/API/CSSStyleRule) calls it this way, also it makes really sense. 

> selectorText is also prefixed with some strange dot, why

Selector is always a selector, a class selector is for e.g. ".my-class-selector" ... you are mixing up the class name and selector.

>  but you can't pass to html class attribute, you can't pass it to some selector engine and you can't produce valid css with it - does nothing. All that when named: false.

Enduser is not using  rule.className nor rule.selector directly, public  api is `sheet.classes``

However I understand your confusion, named and unnamed rules should have a better separation.
The `named` switch makes some logical decisions which might be not obvious for the reader.
I will try to make it better in the next versions. 

But still those are internals, it doesn't make public api for the end user bad.

Also what you probably noticed, `sheet.classes` is empty when named: false. This is because we don't have clear class names. User may  have defined any selector as a key. In this case we can't and we don't need to provide rule.className or sheet.classes consistently with named: true.

I also dislike this inconsistency, but I can't change it. It seems to me the natural consequence of having this option. Not having this option at all is not really an option, because there are valid reasons for having it.

**Comment by kof** [comment]
Created this issue #228

**Comment by kof** [comment]
Also created #229

**Comment by tunnckoCore** [comment]
> Selector is always a selector, a class selector is for e.g. ".my-class-selector" ... you are mixing up the class name and selector.

I understand a bit these OOP things, but.. how this `.div2 > a:hover--jss-0-0` make sense in any kind of context? Because this returns me the `toString` thing, which I believe is my final ready css?

**Comment by kof** [comment]
> I understand a bit these OOP things, but.. how this .div2 > a:hover--jss-0-0 make sense in any kind of context/case? Because this returns me the toString thing, which I believe is my final ready css?

How did you get this selector? Looks like a bug

**Comment by tunnckoCore** [comment]
```js
var jss = require('jss')
var style = {
  'div1 > a:hover': {
    color: 'red'
  }
}

var sheet = jss.createStyleSheet(style)
console.log(sheet)
console.log(sheet.toString())
// .div1 > a:hover--jss-0-0 {
//   color: red;
// }
console.log(sheet.classes) // => { 'div1 > a:hover': 'div1 > a:hover--jss-0-0' }
```

**Comment by kof** [comment]
Oh thats correct, you use a selector as a key, but don't pass named: false option.

**Comment by tunnckoCore** [comment]
Huh.. yea. Okey. In anyway it does to much and do too much magics behind the scenes, which leads to "learn the api / read the docs, read whatever" - it's not natural and intuitive.

Nevermind :) We'll stay tuned. :+1:

**Comment by kof** [comment]
I am open to see ways to solve it better.

**Comment by trusktr** [comment]
> sheet.toString()

Maybe that needs to at least return the proper string.

**Comment by kof** [comment]
@trusktr what do you mean by proper string? how is it not proper right now?

**Comment by trusktr** [comment]
@kof Based on @tunnckoCore's example,

```js
console.log(sheet.toString())
// .div1 > a:hover--jss-0-0 {
//   color: red;
// }
```

that doesn't seem like a usable string that one might expect from `toString`. They might expect

```js
console.log(sheet.toString())
// .div1 > a:hover {
//   color: red;
// }
```

**Comment by kof** [comment]
@trusktr I assume you didn't read the conversation. The result of .toString() is as documented in this case. It's wrong API usage. Whether API is intuitive or not is a different question.

**Comment by trusktr** [comment]
@kof I saw that. Was just pointing out it might be good to consider making that part intuitive (not returning the internal JSS suffix) when `named: false`. Most APIs don't ever make the user think of using `toString`, but, since CSS stylesheets are strings, it might occur to someone to use `toString` for some reason. 

Oh well, it doesn't bother me, as I use your public API. :]

**Comment by kof** [comment]
It is not about toString not being undocumented. It is documented and totally ok to be used. I don't understand your point. If you know how to make it better, please make a concrete proposal in a separate issue.

**Comment by kof** [comment]
The problem is he used key as a selector and didn't pass named: false to the options, so jss used his key as a name not as a selector. See #229

**Comment by trusktr** [comment]
@kof Sorry, yep. Makes sense. I was thinking `named:false` was used, I misread. This works just fine:

```js
import jss from "/client/common/jss-configured"
console.log(
    jss.createStyleSheet({
        '.foo:hover': {
            background: 'red',
        }
    }, {named:false}).toString()
)
// .foo:hover {
//   background: red;
// }
```

@tunnckoCore The output you're seeing *without named:false*, is fine. How can we expect a proper stylesheet in that case? An option, which I don't think is necessary, would be to detect named vs unnamed properties, but that makes selectors like `a` ambiguous. There'd need to be some special syntax for it.

```js
jss.createStyleSheet({
 a: { // named
  width: 1px;
 },
 '-a': { // not named?
  width: 2px;
 }
})
```

**Comment by kof** [comment]
The point of @tunnckoCore is that this option named is confusing, because it lets you use props as names as well as selectors. I am not 100% convinced this is a prob for the user, but it def. makes the logic inside of the Rule less nice, thats why I think it still makes sense to extract the scoping feature into a plugin.

**Comment by nathanmarks** [comment]
Just sharing my 2 cents:

As a user implementing a custom solution with `jss`, I found it difficult at times to navigate the API and bend it to my will.

**Example:**
I'm managing the class name generation myself, and needed the ability to access the `rule` object **before** the [`registerRule()` call here](https://github.com/jsstyles/jss/blob/master/src/StyleSheet.js#L167) in order to have the `classes` property populate on the `StyleSheet` object. Instead -- I've had to proxy the class names manually using my own object.
